main.go

```diff
@@ -267,6 +298,117 @@ func getWebAddress(l net.Listener) string {
 	}
 }
 
+// tryDecodeMultihash attempts to decode input as a multihash in base58 or hex format.
+func tryDecodeMultihash(input string) (cid.Cid, bool) {
+	if mh, err := multihash.FromB58String(input); err == nil {
+		return cid.NewCidV1(cid.Raw, mh), true
+	}
+	if mh, err := multihash.FromHexString(input); err == nil {
+		return cid.NewCidV1(cid.Raw, mh), true
+	}
+	return cid.Cid{}, false
+}
+
+// buildDiagnosticURL returns the appropriate diagnostic URL for a given IPNS name or DNSLink.
+func buildDiagnosticURL(name string) string {
+	if strings.Contains(name, ".") {
+		return "https://dnslink.dev/#" + name
+	}
+	return "https://ipns.ipfs.network/#" + name
+}
+
+// resolveInput attempts to parse input as a CID (with multihash fallback),
+// and if that fails, tries to resolve it as an IPNS name or DNSLink.
+// Returns the resolved CID and optional resolution info.
+func resolveInput(ctx context.Context, ns namesys.NameSystem, input string) (cid.Cid, *MutableResolution, error) {
+	// Strip ipfs:// prefix if present
+	input = strings.TrimPrefix(input, "ipfs://")
+
+	// Try to decode as CID
+	if c, err := cid.Decode(input); err == nil {
+		if c.Type() == libp2pKeyCodec {
+			// PeerID in CIDv1 format - resolve as IPNS without warning
+			return resolveMutablePath(ctx, ns, "/ipns/"+input, false)
+		}
+		// Regular content CID - return immediately
+		return c, nil, nil
+	}
+
+	// Try to decode as legacy PeerID (base58btc multihash)
+	if _, err := peer.Decode(input); err == nil {
+		// Legacy PeerID - resolve as IPNS with warning
+		return resolveMutablePath(ctx, ns, "/ipns/"+input, true)
+	}
+
+	// Try multihash decode as fallback
+	if c, ok := tryDecodeMultihash(input); ok {
+		return c, nil, nil
+	}
+
+	// Must be DNSLink or IPNS path - resolve with warning
+	return resolveMutablePath(ctx, ns, input, true)
+}
+
+// resolveMutablePath resolves an IPNS or DNSLink path to a CID and returns resolution metadata.
+func resolveMutablePath(ctx context.Context, ns namesys.NameSystem, input string, isMutableInput bool) (cid.Cid, *MutableResolution, error) {
+	mutableRes := &MutableResolution{
+		IsMutableInput: isMutableInput,
+	}
+
+	// Normalize input to an IPNS path
+	var p path.Path
+	var err error
+
+	if strings.HasPrefix(input, "/ipns/") || strings.HasPrefix(input, "/ipfs/") {
+		mutableRes.InputPath = input
+		p, err = path.NewPath(input)
+	} else {
+		// Bare domain or IPNS name - prefix with /ipns/
+		mutableRes.InputPath = "/ipns/" + input
+		p, err = path.NewPath(mutableRes.InputPath)
+	}
+
+	if err != nil {
+		return cid.Cid{}, nil, fmt.Errorf("not a valid CID, IPNS name, or DNSLink: %w", err)
+	}
+
+	// If it's an /ipfs/ path, extract the CID directly
+	if !p.Mutable() {
+		segments := p.Segments()
+		c, err := cid.Decode(segments[1])
+		if err != nil {
+			return cid.Cid{}, nil, fmt.Errorf("invalid CID in path: %w", err)
+		}
+		return c, nil, nil
+	}
+
+	// Attempt IPNS/DNSLink resolution
+	result, err := ns.Resolve(ctx, p)
```

Comment:
Perhaps fine for now, but this does full resolution rather than step-wise which could be confusing (e.g. for people who do DNSLink -> IPNS -> IPFS mappings). Should probably have the UI text indicate this is fully resolved rather than partial (again maybe fine, similar to how following CNAMEs, HTTP redirects, etc. is implied in lots of tools)